pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotImplicitTableHintRule.java

```diff
@@ -0,0 +1,283 @@
+        .findAny()
+        .orElse(null);
+  }
+
+  /**
+   * Returns a new node which is a copy of the given table scan with the new table options hint.
+   */
+  private static RelNode withNewTableOptions(LogicalTableScan tableScan, TableOptions tableOptions) {
+    ArrayList<RelHint> newHints = new ArrayList<>(tableScan.getHints());
+
+    newHints.removeIf(relHint -> relHint.hintName.equals(PinotHintOptions.TABLE_HINT_OPTIONS));
+
+    RelHint tableOptionsHint = RelHint.builder(PinotHintOptions.TABLE_HINT_OPTIONS)
+        .hintOption(PinotHintOptions.TableHintOptions.PARTITION_KEY, tableOptions.getPartitionKey())
+        .hintOption(PinotHintOptions.TableHintOptions.PARTITION_FUNCTION, tableOptions.getPartitionFunction())
+        .hintOption(PinotHintOptions.TableHintOptions.PARTITION_SIZE, String.valueOf(tableOptions.getPartitionSize()))
+        .hintOption(PinotHintOptions.TableHintOptions.PARTITION_PARALLELISM,
+            String.valueOf(tableOptions.getPartitionParallelism()))
+        .build();
+
+    newHints.add(tableOptionsHint);
+
+    return tableScan.withHints(newHints);
+  }
+
+  /**
+   * Creates a new table options hint based on the given table partition info and the explicit hint, if any.
+   *
+   * Any explicit hint will override the implicit hint obtained from the table partition info.
+   */
+  private static TableOptions calculateTableOptions(
+      @Nullable RelHint relHint, TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan) {
+    if (relHint == null) {
+      return ImmutablePinotImplicitTableHintRule.TableOptions.builder()
+          .partitionKey(tablePartitionInfo.getPartitionColumn())
+          .partitionFunction(tablePartitionInfo.getPartitionFunctionName())
+          .partitionSize(tablePartitionInfo.getNumPartitions()).partitionParallelism(1)
+          .build();
+    }
+
+    // there is a hint, check fill default data and obtain the partition parallelism if supplied
+    Map<String, String> kvOptions = relHint.kvOptions;
+    return ImmutablePinotImplicitTableHintRule.TableOptions.builder()
+        .partitionKey(getPartitionKey(tablePartitionInfo, tableScan, kvOptions))
+        .partitionFunction(getPartitionFunction(tablePartitionInfo, tableScan, kvOptions))
+        .partitionSize(getPartitionSize(tablePartitionInfo, tableScan, kvOptions))
+        .partitionParallelism(getPartitionParallelism(tableScan, kvOptions))
+        .build();
+  }
+
+  /**
+   * Get the partition key from the hint, if any, otherwise use the partition column from the table partition info.
+   */
+  private static String getPartitionKey(TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan,
+      Map<String, String> kvOptions) {
+    String partitionKey = kvOptions.get(kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_KEY));
+    if (partitionKey != null && !partitionKey.equals(tablePartitionInfo.getPartitionColumn())) {
```

Comment:
Can you explain the need for the second clause in the if condition  `!partitionKey.equals(tablePartitionInfo.getPartitionColumn()))` ?

Isnt this sufficient  ?

```
if (kvOptions.hasKey(...)) {
  return ...
}
```

Reply:
The function would return the same in both cases, but it will log unnecessarily. The idea here is to log if we are using the hint to ask for a partition that is not the one in the configuration. If we use the hint to specify the same partition we would use without the hint no log is necessary. This is specially important because people using colocated joins right now are actually providing the hint with this value, so we don't want to log in that case.

Obviously the log is in debug and won't be actually loged in most environments, but given this is something that is logged for each query I wanted to be conservative.

Reply:
If this is explicitly set to some other value, will it always throw exception in `WorkerManager.getTablePartitionInfo()`?

Reply:
> If this is explicitly set to some other value, will it always throw exception in WorkerManager.getTablePartitionInfo()?

Yes. It may be interesting to remove that in a future PR to be able to use _unsafe_ partitions.